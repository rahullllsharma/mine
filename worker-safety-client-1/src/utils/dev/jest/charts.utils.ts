import type { RiskLevelByDate } from "@/components/charts/dailyRiskHeatmap/types";
import { act } from "@testing-library/react";
import { RiskLevel } from "@/components/riskBadge/RiskLevel";

export function chartRenderHelper(): () => void {
  // set up a listener so we can force the width/height
  // this is required to get the chart to render in its ResponsiveWrapper
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let listener: ((rect: any) => void) | undefined;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (window as any).ResizeObserver = class ResizeObserver {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(ls: any) {
      listener = ls;
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    observe() {}
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    disconnect() {}
  };

  function triggerResize(): void {
    act(() => {
      listener &&
        listener([
          {
            contentRect: {
              x: 0,
              y: 0,
              width: 800,
              height: 800,
              top: 100,
              bottom: 0,
              left: 100,
              right: 0,
            },
          },
        ]);
    });
  }
  return triggerResize;
}

export const sampleHeatmapDates = {
  startDate: "2022-01-29",
  endDate: "2022-02-03",
};

// consumed by the DailyRiskHeatmap
export const sampleRiskLevelByDate: RiskLevelByDate[] = [
  {
    date: "2022-01-29",
    riskLevel: RiskLevel.HIGH,
  },
  {
    date: "2022-01-30",
    riskLevel: RiskLevel.MEDIUM,
  },
  {
    date: "2022-01-31",
    riskLevel: RiskLevel.UNKNOWN,
  },
  {
    date: "2022-02-01",
    riskLevel: RiskLevel.LOW,
  },
  {
    date: "2022-02-02",
    riskLevel: RiskLevel.RECALCULATING,
  },
  {
    date: "2022-02-03",
    riskLevel: RiskLevel.UNKNOWN,
  },
];
