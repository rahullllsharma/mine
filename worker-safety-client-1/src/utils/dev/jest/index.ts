export * from "./intl.utils";
export * from "./testing.utils";
export * from "./charts.utils";
