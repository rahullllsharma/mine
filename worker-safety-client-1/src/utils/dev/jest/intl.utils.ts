/** Mock the default locale when using the Intl.DateTimeFormat class */
export function mockIntlDateTimeFormatLocale(locale: string): void {
  const { DateTimeFormat } = Intl;
  jest
    .spyOn(global.Intl, "DateTimeFormat")
    .mockImplementation((_, options) => new DateTimeFormat(locale, options));
}
