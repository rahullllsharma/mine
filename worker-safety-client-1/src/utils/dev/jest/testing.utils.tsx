/* istanbul ignore file */
import type { ReactNode } from "react";
import type { PropsWithClassName } from "@/types/Generic";
import { fireEvent, render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import { useTenantStore } from "@/store/tenant/useTenantStore.store";
import { TenantMock } from "@/store/tenant/utils/tenantMock";

export const formTemplate = (
  component: ReactNode,
  defaultValues: { [key: string]: any } = {}
) => {
  function Wrapper({ children }: PropsWithClassName): JSX.Element {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  }
  return <Wrapper>{component}</Wrapper>;
};

export const formRender = (
  component: ReactNode,
  defaultValues: { [key: string]: any } = {}
) => {
  return render(formTemplate(component, defaultValues));
};

// to overcome react-select limitation for testing, of not responding to click events
export const openSelectMenu = (role = "combobox", name?: RegExp): void => {
  const inputElement = screen.getByRole(role, { name });
  fireEvent.keyDown(inputElement, { key: "ArrowDown", code: "ArrowDown" });
};

export const mockTenantStore = (mock = TenantMock) => {
  const { setTenant } = useTenantStore.getState();
  setTenant(mock);
};
