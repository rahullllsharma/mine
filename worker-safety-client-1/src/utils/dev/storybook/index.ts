export * from "./WrapperForm.utils";
export * from "./WrapperModal.utils";
export * from "./WrapperScroll.utils";
