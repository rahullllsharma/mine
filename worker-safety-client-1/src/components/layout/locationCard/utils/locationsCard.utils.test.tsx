import type { Activity } from "@/types/activity/Activity";
import { render } from "@testing-library/react";
import {
  TenantMock,
  editTenantEntitiesAttributesMock,
} from "@/store/tenant/utils/tenantMock";
import { mockTenantStore } from "@/utils/dev/jest";
import {
  getCardOptionalProps,
  getActivitiesValues,
} from "./locationsCard.utils";

jest.mock("@/hooks/useTenantFeatures", () => ({
  useTenantFeatures: jest.fn(() => ({
    displayLocationCardDynamicProps: {
      identifier: "activity",
      slot3: "activity",
    },
  })),
}));

describe("Location card utils", () => {
  describe(getActivitiesValues.name, () => {
    beforeAll(() => {
      mockTenantStore(TenantMock);
    });

    describe("when passing empty arrays in the arguments", () => {
      it("should return the default initial object", () => {
        const INPUT = [
          {
            supervisors: [],
          },
        ];
        const OUTPUT = {
          activityTypes: [],
          activitySupervisors: [],
        };

        expect(getActivitiesValues(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing activity type", () => {
      it("should return an object with activities array populated", () => {
        const INPUT = [
          {
            libraryActivityType: {
              id: "1000",
              name: "activity type 1",
            },
            supervisors: [],
          },
        ];
        const OUTPUT = {
          activityTypes: ["activity type 1"],
          activitySupervisors: [],
        };

        expect(getActivitiesValues(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing supervisor", () => {
      it("should return an object with supervisors array populated", () => {
        const INPUT = [
          {
            supervisors: [
              {
                id: "1",
                name: "supervisor 1",
              },
            ],
          },
        ];
        const OUTPUT = {
          activityTypes: [],
          activitySupervisors: ["supervisor 1"],
        };

        expect(getActivitiesValues(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing mulitple activities with multiple activity types and supervisors", () => {
      it("should return an object with both activities and supervisors arrays populated", () => {
        const INPUT = [
          {
            libraryActivityType: {
              id: "1000",
              name: "activity type 1",
            },
            supervisors: [
              {
                id: "1",
                name: "supervisor 1",
              },
              {
                id: "2",
                name: "supervisor 2",
              },
            ],
          },
          {
            libraryActivityType: {
              id: "2000",
              name: "activity type 2",
            },
            supervisors: [
              {
                id: "3",
                name: "supervisor 3",
              },
              {
                id: "4",
                name: "supervisor 4",
              },
            ],
          },
        ];
        const OUTPUT = {
          activityTypes: ["activity type 1", "activity type 2"],
          activitySupervisors: [
            "supervisor 1",
            "supervisor 2",
            "supervisor 3",
            "supervisor 4",
          ],
        };

        expect(getActivitiesValues(INPUT)).toEqual(OUTPUT);
      });
    });
  });

  describe(getCardOptionalProps.name, () => {
    beforeAll(() => {
      mockTenantStore(TenantMock);
    });

    describe("when passing an empty props object", () => {
      it("should return the default object", () => {
        const INPUT = {
          displayLocationCardDynamicProps: {
            identifier: "activity",
            slot3: "activity",
          },
        };
        const OUTPUT = { identifier: "", slots: [] };

        expect(getCardOptionalProps(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing activity slots that have mixed visibility rules in the attributes configuration", () => {
      beforeEach(() => {
        const newMock = editTenantEntitiesAttributesMock([
          {
            entity: "workPackage",
            attribute: "division",
            key: "visible",
            value: false,
          },
        ]);
        mockTenantStore(newMock);
      });

      it("should display only the visible slots", () => {
        const INPUT = {
          division: "division",
          region: "region",
          activities: [
            {
              libraryActivityType: {
                id: "100",
                name: "activity type 1",
              },
              supervisors: [
                {
                  id: "10000",
                  name: "primary assigned person 1",
                },
              ],
            },
          ] as unknown as Activity[],
          displayLocationCardDynamicProps: {
            identifier: "activity",
            slot3: "activity",
          },
        };
        const OUTPUT = {
          identifier: "activity type 1",
          slots: ["region", "primary assigned person 1"],
        };

        expect(getCardOptionalProps(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing location slots that have mixed visibility rules in the attributes configuration", () => {
      beforeEach(() => {
        const newMock = editTenantEntitiesAttributesMock([
          {
            entity: "workPackage",
            attribute: "division",
            key: "visible",
            value: true,
          },
        ]);
        mockTenantStore(newMock);
      });

      it("should display only the visible slots", () => {
        const INPUT = {
          division: "division",
          region: "region",
          primaryAssignedPersonLocation: "primary assigned person",
          displayLocationCardDynamicProps: {
            identifier: "workPackage",
            slot3: "location",
          },
        };
        const OUTPUT = {
          identifier: "",
          slots: ["division", "region", "primary assigned person"],
        };

        expect(getCardOptionalProps(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing workPackage identifier with visibility", () => {
      it("should return the identifier", () => {
        const INPUT = {
          workPackageType: "work package type",
          displayLocationCardDynamicProps: {
            identifier: "workPackage",
            slot3: "location",
          },
        };
        const OUTPUT = {
          identifier: "work package type",
          slots: [],
        };

        expect(getCardOptionalProps(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing workPackageType identifier with no visibility", () => {
      beforeEach(() => {
        const newMock = editTenantEntitiesAttributesMock([
          {
            entity: "workPackage",
            attribute: "workPackageType",
            key: "visible",
            value: false,
          },
        ]);
        mockTenantStore(newMock);
      });

      it("should not return the identifier", () => {
        const INPUT = {
          workPackageType: "work package type",
          displayLocationCardDynamicProps: {
            identifier: "workPackage",
            slot3: "location",
          },
        };
        const OUTPUT = {
          identifier: "",
          slots: [],
        };

        expect(getCardOptionalProps(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing activity identifier with one value", () => {
      it("should return the identifier as a string", () => {
        const INPUT = {
          activities: [
            {
              libraryActivityType: {
                id: "100",
                name: "activity type 1",
              },
              supervisors: [],
            },
          ] as unknown as Activity[],
          displayLocationCardDynamicProps: {
            identifier: "activity",
            slot3: "activity",
          },
        };
        const OUTPUT = {
          identifier: "activity type 1",
          slots: [],
        };

        expect(getCardOptionalProps(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing activity identifier with multiple values", () => {
      it("should return the identifier as an element", () => {
        const { identifier } = getCardOptionalProps({
          activities: [
            {
              libraryActivityType: {
                id: "100",
                name: "activity type 1",
              },
              supervisors: [],
            },
            {
              libraryActivityType: {
                id: "200",
                name: "activity type 2",
              },
              supervisors: [],
            },
          ] as unknown as Activity[],
          displayLocationCardDynamicProps: {
            identifier: "activity",
            slot3: "activity",
          },
        });
        expect(typeof identifier).toBe("object");
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const { asFragment } = render(identifier);
        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe("when passing one activity with multiple supersivor", () => {
      it("should return the supervisor in the last slot", () => {
        const INPUT = {
          activities: [
            {
              libraryActivityType: null,
              supervisors: [
                {
                  id: "10000",
                  name: "primary assigned person 1",
                },
              ],
            },
          ] as unknown as Activity[],
          displayLocationCardDynamicProps: {
            identifier: "activity",
            slot3: "activity",
          },
        };
        const OUTPUT = {
          identifier: "",
          slots: ["primary assigned person 1"],
        };

        expect(getCardOptionalProps(INPUT)).toEqual(OUTPUT);
      });
    });

    describe("when passing multiple activities with multiple supersivors", () => {
      it("should return the supervisors in the last slot", () => {
        const INPUT = {
          activities: [
            {
              libraryActivityType: null,
              supervisors: [
                {
                  id: "10000",
                  name: "primary assigned person 1",
                },
                {
                  id: "20000",
                  name: "primary assigned person 2",
                },
              ],
            },
            {
              libraryActivityType: null,
              supervisors: [
                {
                  id: "30000",
                  name: "primary assigned person 3",
                },
                {
                  id: "40000",
                  name: "primary assigned person 4",
                },
              ],
            },
          ] as unknown as Activity[],
          displayLocationCardDynamicProps: {
            identifier: "activity",
            slot3: "activity",
          },
        };
        const OUTPUT = {
          identifier: "",
          slots: [
            "primary assigned person 1, primary assigned person 2, primary assigned person 3, primary assigned person 4",
          ],
        };

        expect(getCardOptionalProps(INPUT)).toEqual(OUTPUT);
      });
    });
  });
});
