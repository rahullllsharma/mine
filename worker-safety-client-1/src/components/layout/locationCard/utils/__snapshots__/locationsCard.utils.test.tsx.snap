// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location card utils getCardOptionalProps when passing activity identifier with multiple values should return the identifier as an element 1`] = `
<DocumentFragment>
  <span
    class="inline-flex items-center"
  >
    <span
      class="inline-flex items-center w-full"
    >
      Multiple
    </span>
  </span>
</DocumentFragment>
`;
